### PR TITLE
RUM-2154 Enable resource endpoint

### PR DESCRIPTION
**Pull request by maciejburda**
### What and why?

Enables resource recording using resource endpoint and deprecates and disables base64 solution.
Along with this, we conducted some E2E tests and fixed some minor bugs that we spotted.

Also, there were a few updates to snapshot tests, that cover new ways of getting and rendering image data.

### How?

Code was prepared in past PRs. It's mostly achieved by injecting `ResourceProcessor`.

### Review checklist
- [ ] Feature or bugfix MUST have appropriate tests (unit, integration)
- [ ] Make sure each commit and the PR mention the Issue number or JIRA reference
- [ ] Add CHANGELOG entry for user facing changes

### Custom CI job configuration (optional)
- [ ] Run unit tests for Core, RUM, Trace, Logs, CR and WVT
- [ ] Run unit tests for Session Replay
- [ ] Run integration tests
- [ ] Run smoke tests
- [ ] Run tests for `tools/`

### CODE REVIEW

**Comment by maciejburda** [comment]
To be rebased on develop after https://github.com/DataDog/dd-sdk-ios/pull/1585 is merged

**Comment by maciejburda** [comment]
Awaiting for backend to be deployed.

File: DatadogSessionReplay/Sources/Recorder/Utilities/UIImage+SessionReplay.swift [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+ * Copyright 2019-Present Datadog, Inc.
+ */
+
+#if os(iOS)
+import Foundation
+import UIKit
+import CryptoKit
+
+private var srIdentifierKey: UInt8 = 11
+private var recordedKey: UInt8 = 22
```
Comment (maciejburda) [inline DatadogSessionReplay/Sources/Recorder/Utilities/UIImage+SessionReplay.swift:13]: remove

File: DatadogSessionReplay/Sources/Processor/ResourcesProcessor.swift [hunk 2]
```diff
@@ -13,22 +13,40 @@ internal protocol ResourceProcessing {
...
                     EnrichedResource(
                         identifier: $0.calculateIdentifier(),
                         data: $0.calculateData(),
                         context: context
                     )
                 }
+                .filter {
+                    let isIncluded = self?.processedIdentifiers.contains($0.identifier) == false
+                    self?.processedIdentifiers.insert($0.identifier)
+                    return isIncluded
```
Comment (maciejburda) [inline DatadogSessionReplay/Sources/Processor/ResourcesProcessor.swift:43]: improve insert condition

File: DatadogSessionReplay/Sources/Recorder/Utilities/UIImage+SessionReplay.swift [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+extension UIImage {
+    var srIdentifier: String {
```
Comment (maciejburda) [inline DatadogSessionReplay/Sources/Recorder/Utilities/UIImage+SessionReplay.swift:15]: add thread sync

**Review by ncreated — COMMENTED** [review]
I haven't looked SR product changes, but modifications in `sr-snapshots` CLI and `SRSnapshotTests` project look fine with one remark 👌. Glad we extended it with the support to SR Resources - nice 💪.

**Review by maxep — CHANGES_REQUESTED** [review]
Looks good 👍  I have left a couple of blocking points

File: DatadogSessionReplay/Sources/Processor/ResourcesProcessor.swift [hunk 4]
```diff
@@ -13,23 +13,43 @@ internal protocol ResourceProcessing {
 }
 
 internal class ResourceProcessor: ResourceProcessing {
+    #if DEBUG
+    /// Interception callback for snapshot tests.
+    /// Only available in Debug configuration, solely made for testing purpose.
+    var interceptResources: (([Resource]) -> Void)? = nil
+    #endif
+
```
[the remark this answers is not in the document]
Comment (maciejburda) [reply DatadogSessionReplay/Sources/Processor/ResourcesProcessor.swift:21]: No strong opinion on this. I've done it similarly to `SnapshotProcessor`. I'll remove it in both.

File: DatadogSessionReplay/Sources/Processor/ResourcesProcessor.swift [hunk 5]
```diff
@@ -13,23 +13,43 @@ internal protocol ResourceProcessing {
...
-                    EnrichedResource(
-                        identifier: $0.calculateIdentifier(),
-                        data: $0.calculateData(),
-                        context: context
-                    )
+        #if DEBUG
+        interceptResources?(resources)
+        #endif
+        queue.run { [weak self] in
+            autoreleasepool {
```
[the remark this answers is not in the document]
Comment (maciejburda) [reply DatadogSessionReplay/Sources/Processor/ResourcesProcessor.swift:32]: I initially did that, but later on decided to narrow down the scope to limit the impact. But it makes sense to change it globally in the `BackgroundAsyncQueue` that holds the `DispatchQueue`.

**Comment by maciejburda** [comment]
Thanks for the review @maxep

I think I managed to resolve all the comments.

Also simplified threading by reusing the same queue in both processors.

File: DatadogSessionReplay/Sources/Recorder/Utilities/UIImage+SessionReplay.swift [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+#if os(iOS)
+import Foundation
+import UIKit
+import DatadogInternal
+import CryptoKit
+
+private var srIdentifierKey: UInt8 = 11
+extension UIImage: DatadogExtended {}
+extension DatadogExtension where ExtendedType: UIImage {
+    var srIdentifier: String {
```
Comment (ncreated) [inline DatadogSessionReplay/Sources/Recorder/Utilities/UIImage+SessionReplay.swift:16]: suggestion/ add tests for `srIdentifier`

File: DatadogSessionReplay/Sources/Recorder/ViewTreeSnapshotProducer/ViewTreeSnapshot/NodeRecorders/UIImageResource.swift [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+        guard let tintColor = tintColor else {
+            return image.scaledDownToApproximateSize(SessionReplay.maxObjectSize)
+        }
+        if #available(iOS 13.0, *), image.isSymbolImage {
+            return image.withTintColor(tintColor)
+                .scaledDownToApproximateSize(SessionReplay.maxObjectSize)
+        } else {
+            return manuallyTintedImageData()
+        }
+    }
```
Comment (ncreated) [inline DatadogSessionReplay/Sources/Recorder/ViewTreeSnapshotProducer/ViewTreeSnapshot/NodeRecorders/UIImageResource.swift:39]: suggestion/ add tests for identifier and data calculation

**Review by ncreated — APPROVED** [review]
Looks ok 👌, I left suggestions on adding tests coverage

File: DatadogSessionReplay/Sources/Recorder/Utilities/UIImage+SessionReplay.swift [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+        srIdentifierLock.lock()
+        defer { srIdentifierLock.unlock() }
+        if let hash = objc_getAssociatedObject(self, &srIdentifierKey) as? String {
+            return hash
+        } else {
+            let hash = computeHash()
+            objc_setAssociatedObject(self, &srIdentifierKey, hash, .OBJC_ASSOCIATION_RETAIN_NONATOMIC)
+            return hash
+        }
```
[the remark this answers is not in the document]
Comment (maxep) [reply DatadogSessionReplay/Sources/Recorder/Utilities/UIImage+SessionReplay.swift]: @maciejburda we are removing atomicity of the `computeHash` access? 🤔

[the remark this answers is not in the document]
Comment (maciejburda) [reply DatadogSessionReplay/Sources/Recorder/Utilities/UIImage+SessionReplay.swift]: yes - I refactored to just one queue. Had issues making it work last time

**Review by maxep — APPROVED** [review]
Looks great 👍

File: DatadogSessionReplay/Sources/Recorder/Utilities/UIImage+SessionReplay.swift (hunk 6 above)
Comment (maciejburda) [reply DatadogSessionReplay/Sources/Recorder/Utilities/UIImage+SessionReplay.swift:16]: covered through UIImageResource tests

File: DatadogSessionReplay/Sources/Recorder/ViewTreeSnapshotProducer/ViewTreeSnapshot/NodeRecorders/UIImageResource.swift (hunk 7 above)
Comment (maciejburda) [reply DatadogSessionReplay/Sources/Recorder/ViewTreeSnapshotProducer/ViewTreeSnapshot/NodeRecorders/UIImageResource.swift:39]: Added some tests with soft assertions. Different simulators / iOS versions might produce different images / hashes.

It's more fully tested through snapshot tests that render this calculatedData() as images.

**Review by ncreated — COMMENTED** [review]
🙏 Can we have this PR properly described, not leaving blank PR template? Not sure what is the scope of this change.

**Pull request merged**